consul/prepared_query_endpoint.go

```diff
@@ -368,10 +368,44 @@ func (p *PreparedQuery) Execute(args *structs.PreparedQueryExecuteRequest,
 	// Shuffle the results in case coordinates are not available if they
 	// requested an RTT sort.
 	reply.Nodes.Shuffle()
-	if err := p.srv.sortNodesByDistanceFrom(args.Source, reply.Nodes); err != nil {
+
+	// Get the source to sort by. This can be passed in by the requestor, or
+	// pre-defined using the Near parameter in the prepared query. If the
+	// near parameter was defined, that will be preferred.
+	sortFrom := args.Source
+	if query.Service.Near != "" {
+		sortFrom = structs.QuerySource{
+			Datacenter: args.Datacenter,
+			Node:       query.Service.Near,
+		}
+	}
+
+	// Respect the magic "_agent" flag.
+	preferLocal := false
+	if sortFrom.Node == "_agent" {
+		preferLocal = true
+		sortFrom = args.Origin
+	}
+
+	// Perform the distance sort
+	if err := p.srv.sortNodesByDistanceFrom(sortFrom, reply.Nodes); err != nil {
 		return err
 	}
 
+	// Nodes cannot be any "closer" than localhost, so this special case ensures
+	// the local node is returned first if it is present in the result. This
+	// allows the local agent to be preferred even when network coordinates are
+	// not enabled. Only works if the results come from the request origin DC.
+	if preferLocal && reply.Datacenter == args.Origin.Datacenter {
```

Comment:
I don't think the special case is worth it since 0.6 has been out for a long time. This way you won't need to know about `_agent` at all at this level.